File: backend/model/product.rb
class Product
  # use type to distinguish each kind of product: physical, book, digital, membership, etc.
  attr_reader :name, :type, :price

  def initialize(name:, type:, price:)
    @name, @type, @price = name, type, price
  end

  def actions
    @actions ||= case @type
    when :physical
      PaymentActions.new([AddressShippingLabel.new])
    when :membership
      PaymentActions.new([
        ActiveMembership.new,
        EmailActiveMembership.new
      ])
    when :book
      PaymentActions.new([AddressShippingLabel.new('Isento de impostos.')])
    when :digital
      PaymentActions.new([
        EmailPayment.new,
        CreateVoucher.new(discount = 10)
      ])
    else
      PaymentActions.new([])
    end
  end
end

File: backend/proposal/core/product.py
class Product:
    # use type to distinguish each kind of product: physical, book, digital, membership, etc.
    name = None
    type = None
    AVAILABLE_TYPES=["physical", "book", "digital", "membership"]

    def __init__(self, name, type):
        self.name = name
        # Checking if type is valid
        if type in self.AVAILABLE_TYPES:
            self.type = type
        else:
            # Throwing error
            raise ValueError("Invalid type of product")
